File: src/main/java/com/icube/sim/tichu/games/hanabi/dtos/HanabiMessage.java
package com.icube.sim.tichu.games.hanabi.dtos;

import com.icube.sim.tichu.games.hanabi.HanabiRule;
import lombok.Getter;
import org.jspecify.annotations.Nullable;

@Getter
public class HanabiMessage {
    private final HanabiMessageType type;
    @Nullable
    private final Object data;

    private HanabiMessage(HanabiMessageType type, @Nullable Object data) {
        this.type = type;
        this.data = data;
    }

    public static HanabiMessage setRule(HanabiRule rule) {
        return new HanabiMessage(HanabiMessageType.SET_RULE, rule);
    }

    public static HanabiMessage start() {
        return new HanabiMessage(HanabiMessageType.START, null);
    }

    public static HanabiMessage state(HanabiDto dto) {
        return new HanabiMessage(HanabiMessageType.STATE, dto);
    }

    public static HanabiMessage hint(HintInfo info) {
        return new HanabiMessage(HanabiMessageType.HINT, info);
    }

    public static HanabiMessage discard(DiscardInfo info) {
        return new HanabiMessage(HanabiMessageType.DISCARD, info);
    }

    public static HanabiMessage play(PlayInfo info) {
        return new HanabiMessage(HanabiMessageType.PLAY, info);
    }

    public static HanabiMessage resolveFreeHint(FreeHintInfo info) {
        return new HanabiMessage(HanabiMessageType.RESOLVE_FREE_HINT, info);
    }

    public static HanabiMessage resolveRecoverToDeck(RecoverToDeckInfo info) {
        return new HanabiMessage(HanabiMessageType.RESOLVE_RECOVER_TO_DECK, info);
    }

    public static HanabiMessage resolveRecoverToPlay(RecoverToPlayInfo info) {
        return new HanabiMessage(HanabiMessageType.RESOLVE_RECOVER_TO_PLAY, info);
    }

    public static HanabiMessage end(HanabiDto dto) {
        return new HanabiMessage(HanabiMessageType.END, dto);
    }
}

File: frontend/src/games/hanabi/dtos/HanabiMessage.ts
import { BonusEffect, ClueType, HanabiCardDto, HanabiColor } from "@/games/hanabi/dtos/HanabiDto.ts";

export enum HanabiMessageType {
  SET_RULE = 'SET_RULE',
  START = 'START',
  STATE = 'STATE',
  HINT = 'HINT',
  DISCARD = 'DISCARD',
  PLAY = 'PLAY',
  RESOLVE_FREE_HINT = 'RESOLVE_FREE_HINT',
  RESOLVE_RECOVER_TO_DECK = 'RESOLVE_RECOVER_TO_DECK',
  RESOLVE_RECOVER_TO_PLAY = 'RESOLVE_RECOVER_TO_PLAY',
  END = 'END',
}

export interface HanabiMessage {
  type: HanabiMessageType;
  data: unknown;
}

// ----- Per-action delta payloads (carried by the messages above) -----
// These mirror the backend `*Info` records. They describe only what the action changed; the resulting
// scalar counters (tokens, deck size, turn) are derived client-side by `applyDelta`. A `drawnCard` that
// is present but masked (`{color: null, value: null}`) still means a card was drawn.

export interface HintInfo {
  fromId: number;
  targetId: number;
  clueType: ClueType;
  color: HanabiColor | null;
  value: number | null;
  matchedIndices: number[];
}

export interface DiscardInfo {
  playerId: number;
  index: number;
  discardedCard: HanabiCardDto;
  drawnCard: HanabiCardDto | null;
}

export interface PlayInfo {
  playerId: number;
  index: number;
  playedCard: HanabiCardDto;
  success: boolean;
  fireworkCompleted: boolean;
  bonusEffect: BonusEffect | null;
  drawnCard: HanabiCardDto | null;
}

export interface FreeHintInfo {
  fromId: number;
  targetId: number;
  clueType: ClueType;
  color: HanabiColor | null;
  value: number | null;
  matchedIndices: number[];
  drawnCard: HanabiCardDto | null;
}

export interface RecoverToDeckInfo {
  playerId: number;
  recoveredCard: HanabiCardDto;
  drawnCard: HanabiCardDto | null;
}

export interface RecoverToPlayInfo {
  playerId: number;
  index: number;
  playedCard: HanabiCardDto;
  fireworkCompleted: boolean;
  bonusEffect: BonusEffect | null;
  drawnCard: HanabiCardDto | null;
}
